src/uct/ib/ud/base/ud_iface.c

```diff
@@ -658,17 +665,36 @@ static void uct_ud_iface_free_res_skbs(uct_ud_iface_t *iface)
 void uct_ud_iface_dispatch_async_comps_do(uct_ud_iface_t *iface)
 {
     uct_ud_comp_desc_t *cdesc;
-    uct_ud_send_skb_t *skb;
+    uct_ud_send_skb_t  *skb;
+    ucs_queue_head_t eps_failed;
+    uct_ud_ep_t      *ep_failed;
+    ucs_queue_head_init(&eps_failed);
 
     do {
         skb = ucs_queue_pull_elem_non_empty(&iface->tx.async_comp_q,
                                             uct_ud_send_skb_t, queue);
         cdesc = uct_ud_comp_desc(skb);
-        uct_invoke_completion(cdesc->comp, UCS_OK);
-        cdesc->ep->flags &= ~UCT_UD_EP_FLAG_ASYNC_COMPS;
-        skb->flags = 0;
+
+        if (ucs_unlikely(skb->flags & UCT_UD_SEND_SKB_FLAG_ERR)) {
+            if (skb->flags & UCT_UD_SEND_SKB_FLAG_COMP) {
+                uct_invoke_completion(cdesc->comp, skb->neth->status);
+            }
+            cdesc->ep->flags &= ~UCT_UD_EP_FLAG_ASYNC_COMPS;
+            skb->flags = 0;
+            ucs_queue_push_uniq(&eps_failed, cdesc->ep, fail_queue);
```

Comment:
I'd not use queue here to avoid extending UD ep size. For instance we  can invoke uct_set_ep_failed() function here. And we may set some special flag on ep to mark that it's already moved to failed state.